strangerstudios/paid-memberships-pro
# Fix pmpro_search_filter to support custom taxonomies and archives better.

There is an issue where WooCommerce products are incorrectly being filtered out of searches and archive pages, even the product cat archive pages, when the "filter searches and archives" advanced setting is set.

In reality, we want to just filter out restricted members-only content, and allow everything else.

There was a previous fix here (https://github.com/strangerstudios/paid-memberships-pro/pull/2717) but in reviewing that we realized that this whole function/etc is due for refactoring.

We don't want to refactor for the sake of it, but refactoring this will make it possible to fix this but, some related bugs, and to better maintain this code going forward.

It was hard to limit the scope of the refactor. I think we'll end up rewriting a lot of this. We can merge it into the 3.0 release to help minimize frustrations with this code changing. Even though we should end up with something that is "better" and/or something that fixes the filter to function as intended, people have been running sites off this code and may expect things to keep working "the old way". i.e. people will update and may notice that some things that used to be hidden are no longer hidden... some things that used to be visible are no longer visible.

Okay. So far, we've:
* Removed some redundant code to cast the post_types query var to an array.
* Broken out the logic for when the filter runs into individual checks that return early.
* Fixed a hard coded dev_ DB prefix in the where clause filter code.
* Tweaked the post_type check so that the filter may still run even if the post_type query var was empty.
* Added an archive check to skip the filter if we're viewing an archive that is not in the list to be filtered. 

Those last 2 bullet points basically fix the main issue that was noticed. If you navigate to a WooCommerce product category, the products are no longer being filtered (unless you adjust the pmpro_search_filter_post_types filter to include products).

But we have some other issues and things we might want to look at.

Other issues:
* While the current version of the PR fixed archive views, we are still filtering products out of searches even if products aren't in the list of post types to filter.
* Even if products are in the list of post types to filter, the filter isn't working correctly for WC products or probably other CPTs as well.

To address these, we need to look into the code the sets the `category__in` query var. We only look up post categories to set this var, so it's not going to include any taxonomies that aren't for posts. So basically any other CPT in a cat gets excluded.

(Note: We want to tests posts and products without a query too. These should show up in the search if they aren't otherwise restricted.)

We maybe can switch to using `category__not_in` instead, but it seems like we might have had a reason to do it the other way. We should probably rethink this from scratch.

Other things to look into:
* This code looks like it is potentially making at least 1 additional query for every level, and it may not be caching this stuff well if there are multiple queries on page load that require filtering like this. We should try to combine queries and basically make this as performant as possible. It seems like the query to get cats could be one query for all levels instead of individual queries.
* This is all called the *search* filter. Should we filter archives using a totally separate callback? The functions could start with query checks to see if it looks like a search or archive and return early otherwise. Splitting things up may help.
* We are bailing on the filter if the query is a REST API call. Does this mean that private content could potentially leak through REST API calls? Are we otherwise protecting that? Why do we need to bail on REST API queries anyway? Maybe there is a better check.

### All Submissions:






### Changes proposed in this Pull Request:



Resolves: https://github.com/strangerstudios/pmpro-woocommerce/issues/195

### How to test the changes in this Pull Request:

1. Install WC, PMPro
2. Create a couple products. Put them in the same product category.
3. Regular setup of PMPro, NO products are marked to require membership or have membership discounts.
4. Make sure the Memberships > Settings > Advanced Settings > Filter Searches and Archives setting is set to "Yes"
5. Visit the category archive page. Note if 2 products are shown.

We also want to test searches for terms that should turn up the products, but currently aren't (in the first version of the PR).

Test both searches and archive pages. Test custom archives and post category archives. Make sure that posts which should be hidden are hidden and posts that should be visible are visible. This can get confusing. Have some posts that are in cats that are restricted. Have some posts that are restricted by level themselves. Have some posts that aren't restricted by level. Have posts in non-member cats. Have posts in no cat. Restrict a product using custom code or the PMPro CPT add on. Make sure they are hidden when expected.

### Other information:




### Changelog entry

> BUG FIX/ENHANCEMENT: Fixes to the code used to filter members-only content from searches and archives.
